test/TensorFlowRuntime/tensor_group.swift

```diff
@@ -0,0 +1,78 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift -swift-version 5 -g %s -o %t/a.out
+// RUN: %target-codesign %t/a.out
+// RUN: %target-run %t/a.out
+// REQUIRES: executable_test
+//
+// `TensorGroup` tests.
+
+import TensorFlow
+import StdlibUnittest
+
+var TensorGroupTests = TestSuite("TensorGroup")
+
+extension TensorDataType : Equatable {
+  public static func == (lhs: TensorDataType, rhs: TensorDataType) -> Bool {
+    return Int(lhs._cDataType.rawValue) == Int(rhs._cDataType.rawValue)
+  }
+}
+
+struct Empty : TensorGroup {}
+
+struct Simple : TensorGroup {
+  var w, b: Tensor<Float>
+}
+
+struct Mixed : TensorGroup {
+  // Mutable.
+  var string: StringTensor
+  var float: Tensor<Float>
+  // Immutable.
+  let int: Tensor<Int32>
+}
+
+struct Nested : TensorGroup {
+  // Immutable.
+  let simple: Simple
+  // Mutable.
+  var mixed: Mixed
+}
+
+struct Generic<T: TensorGroup, U: TensorGroup> : TensorGroup {
+  var t: T
+  var u: U
```

Comment:
Could you add some tests for nested types, for example:

```swift
struct Foo<T: TenosrGroup> {
  func foo<U: TensorGroup>() {
    struct UltraNested<V: TensorGroup> {
      var a: Generic<T, U>
      var b: Generic<U, V>
      var c: Generic<T, V>
    }
  }
}
```

Same for `TensorArrayProtocol`.

cc @dan-zheng. I remember seeing a crash when a nested type uses a generic parameter from the parent function, when writing the custom differentiation tutorial. Would be great to test this for `Differentiable` derived conformances as well.

Reply:
> cc @dan-zheng. I remember seeing a crash when a nested type uses a generic parameter from the parent function, when writing the custom differentiation tutorial. Would be great to test this for Differentiable derived conformances as well.

Yes, I remember this bug. It's a latent bug that first existed in `Parameterized` derived conformances, then got copied elsewhere.

It shouldn't be hard to fix - it's a matter of calling `mapTypeIntoContext` on the correct `DeclContext`. I can take a look if the bug surfaces.

Reply:
Actually I remember being confused about which `DeclContext` to use in each case, but I thought it through a bit and hope it should not exist now. I'll add some more tests for that.

Reply:
Added a couple simple nested generic tests and they seem to work fine.

Reply:
The test case you added is a bit different from the example I gave. In my example `UltraNested` is defined in the body of a generic function. Types nested in function scopes were places where I've seen crashers.

Reply:
Yes that was intentional because the example above does not compile. The Swift compiler says that I am not allowed to define a nested struct in a generic function.

Reply:
Can you make the function be non-generic?

Reply:
So, even when I do that, how can I refer to that nested struct type from outside the function so that I can test its static properties?

Reply:
Test by calling the function.
```swift
struct Foo<T: TenosrGroup> {
  func foo() {
    struct UltraNested<U: TensorGroup, V: TensorGroup> {
      var a: Generic<T, U>
      var b: Generic<U, V>
      var c: Generic<T, V>
    }
    let x: UltraNested = ...
    expectEqual(...)
  }
}
Foo(...).foo()
```

Reply:
Thanks! I just added some tests that are almost what you want. I had to move `T` inside the function because otherwise I get `error: type 'UltraNested' cannot be nested in generic function 'function()'`. If that's good, is there anything else remaining to do?

Reply:
Looks great!